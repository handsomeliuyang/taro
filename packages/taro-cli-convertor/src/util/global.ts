import { ErrCodeMsg } from './index'

interface Globals {
  logFilePath: string
  errCodeMsgs: ErrCodeMsg[]
  currentParseFile: string
}

export const globals: Globals = {
  logFilePath: '',
  errCodeMsgs: [],
  currentParseFile: ''
}
