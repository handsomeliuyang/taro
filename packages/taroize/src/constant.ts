export const reserveKeyWords = new Set([
  'break',
  'case',
  'catch',
  'continue',
  'default',
  'delete',
  'do',
  'else',
  'finally',
  'for',
  'function',
  'if',
  'in',
  'instanceof',
  'new',
  'return',
  'switch',
  'this',
  'throw',
  'try',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'abstract',
  'boolean',
  'byte',
  'char',
  'class',
  'const',
  'debugger',
  'double',
  'enum',
  'export',
  'extends',
  'final',
  'float',
  'goto',
  'implements',
  'import',
  'int',
  'interface',
  'long',
  'native',
  'package',
  'private',
  'protected',
  'public',
  'short',
  'static',
  'super',
  'synchronized',
  'throws',
  'transient',
])

export const InternalKeyWorld = new Set([
  'Array',
  'Object',
  'String',
  'Number',
  'Boolean',
  'Function',
  'Date',
  'RegExp',
  'Math',
  'Error',
  'JSON',
  'Promise',
  'Map',
  'Symbol',
  'Reflect',
])

export const navigateFunc = new Set(['navigateTo', 'redirectTo', 'reLaunch'])
