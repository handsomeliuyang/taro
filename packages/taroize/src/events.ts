export const specialEvents = new Map<string, string>()
specialEvents.set('bindtimeupdate', 'onTimeUpdate')
specialEvents.set('bindgetphoneNumber', 'onGetPhoneNumber')
specialEvents.set('bindchooseavatar', 'onChooseAvatar')
specialEvents.set('bindgetrealnameauthinfo', 'onGetRealnameAuthInfo')
specialEvents.set('bindopensetting', 'onOpenSetting')
specialEvents.set('bindscancode', 'onScanCode')
specialEvents.set('bindstatechange', 'onStateChange')
specialEvents.set('bindhtouchmove', 'onHTouchMove')
specialEvents.set('bindvtouchmove', 'onVTouchMove')
specialEvents.set('bindcolumnchange', 'onColumnChange')
specialEvents.set('bindscrolltoupper', 'onScrollToUpper')
specialEvents.set('bindscrolltolower', 'onScrollToLower')
specialEvents.set('bindanimationfinish', 'onAnimationFinish')
specialEvents.set('bindfullscreenchange', 'onFullscreenChange')
specialEvents.set('bindtouchstart', 'onTouchStart')
specialEvents.set('bindtouchmove', 'onTouchMove')
specialEvents.set('bindtouchcancel', 'onTouchCancel')
specialEvents.set('bindtouchend', 'onTouchEnd')
specialEvents.set('bindlongpress', 'onLongPress')
specialEvents.set('bindlongclick', 'onLongClick')
specialEvents.set('bindtransitionend', 'onTransitionEnd')
specialEvents.set('bindanimationstart', 'onAnimationStart')
specialEvents.set('bindanimationtteration', 'onAnimationIteration')
specialEvents.set('bindanimationend', 'onAnimationEnd')
specialEvents.set('bindtouchforcechange', 'onTouchForceChange')
specialEvents.set('bindtap', 'onClick')
specialEvents.forEach((value, key) => {
  specialEvents.set(key.replace(/^bind/, 'catch'), value)
})

export const specialSameEvents = new Map<string, string>()
specialSameEvents.set('bind:timeupdate', 'onTimeUpdate')
specialSameEvents.set('bind:getphoneNumber', 'onGetPhoneNumber')
specialSameEvents.set('bind:chooseavatar', 'onChooseAvatar')
specialSameEvents.set('bind:getrealnameauthinfo', 'onGetRealnameAuthInfo')
specialSameEvents.set('bind:opensetting', 'onOpenSetting')
specialSameEvents.set('bind:scancode', 'onScanCode')
specialSameEvents.set('bind:statechange', 'onStateChange')
specialSameEvents.set('bind:htouchmove', 'onHTouchMove')
specialSameEvents.set('bind:vtouchmove', 'onVTouchMove')
specialSameEvents.set('bind:columnchange', 'onColumnChange')
specialSameEvents.set('bind:scrolltoupper', 'onScrollToUpper')
specialSameEvents.set('bind:scrolltolower', 'onScrollToLower')
specialSameEvents.set('bind:animationfinish', 'onAnimationFinish')
specialSameEvents.set('bind:fullscreenchange', 'onFullscreenChange')
specialSameEvents.set('bind:touchstart', 'onTouchStart')
specialSameEvents.set('bind:touchmove', 'onTouchMove')
specialSameEvents.set('bind:touchcancel', 'onTouchCancel')
specialSameEvents.set('bind:touchend', 'onTouchEnd')
specialSameEvents.set('bind:longpress', 'onLongPress')
specialSameEvents.set('bind:longclick', 'onLongClick')
specialSameEvents.set('bind:transitionend', 'onTransitionEnd')
specialSameEvents.set('bind:animationstart', 'onAnimationStart')
specialSameEvents.set('bind:animationtteration', 'onAnimationIteration')
specialSameEvents.set('bind:animationend', 'onAnimationEnd')
specialSameEvents.set('bind:touchforcechange', 'onTouchForceChange')
specialSameEvents.set('bind:tap', 'onClick')
specialSameEvents.forEach((value, key) => {
  specialSameEvents.set(key.replace(/^bind/, 'catch'), value)
})