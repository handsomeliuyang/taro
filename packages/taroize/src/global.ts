export const usedComponents = new Set<string>()

export const errors: string[] = []

export const globals = {
  hasCatchTrue: false,
  logFilePath: '',
}

export const THIRD_PARTY_COMPONENTS = new Set<string>()

export const resetGlobals = (logFilePath) => {
  globals.hasCatchTrue = false
  globals.logFilePath = logFilePath
  THIRD_PARTY_COMPONENTS.clear()
}
