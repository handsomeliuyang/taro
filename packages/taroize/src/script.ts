import traverse, { NodePath, Visitor } from '@babel/traverse'
import * as t from '@babel/types'

import { usedComponents } from './global'
import { buildBlockElement, buildImportStatement, buildRender, codeFrameError, parseCode } from './utils'
import { WXS } from './wxml'

const defaultClassName = '_C'

const buildDecorator = (id: t.Identifier | t.ObjectExpression, isApp = false) => {
  const args: any[] = [id]
  isApp && args.push(t.booleanLiteral(true))
  return t.decorator(t.callExpression(t.identifier('withWeapp'), args))
}

export function replaceIdentifier (callee: NodePath<t.Node>) {
  if (callee.isIdentifier()) {
    const name = callee.node.name
    if (name === 'getApp' || name === 'getCurrentPages') {
      callee.replaceWith(t.memberExpression(t.identifier('Taro'), callee.node))
    }
  }
}

// babel 6升级babel 7适配
export function replaceMemberExpression (callee: NodePath<t.Node>) {
  if (callee.isMemberExpression()) {
    const object = callee.get('object') as NodePath<t.Identifier>
    if (t.isIdentifier(object.node, { name: 'wx' })) {
      object.replaceWith(t.identifier('Taro'))
    }
  }
}

export function parseScript (
  script?: string,
  scriptPath?: string,
  returned?: t.Expression,
  wxses: WXS[] = [],
  refId?: Set<string>,
  isApp = false
) {
  script = script || 'Page({})'
  if (t.isJSXText(returned as any)) {
    const block = buildBlockElement()
    block.children = [returned as any]
    returned = block
  }
  let ast = parseCode(script, scriptPath as string)
  let classDecl!: t.ClassDeclaration
  let foundWXInstance = false
  const vistor: Visitor = {
    BlockStatement (path) {
      path.scope.rename('wx', 'Taro')
    },
    Identifier (path) {
      if (path.isReferenced() && path.node.name === 'wx') {
        path.replaceWith(t.identifier('Taro'))
      }
    },
    CallExpression (path) {
      const callee = path.get('callee')
      replaceIdentifier(callee as NodePath<t.Node>)
      replaceMemberExpression(callee as NodePath<t.Node>)
      if (
        callee.isIdentifier({ name: 'Page' }) ||
        callee.isIdentifier({ name: 'Component' }) ||
        callee.isIdentifier({ name: 'App' })
      ) {
        foundWXInstance = true
        const componentType = callee.node.name
        classDecl = parsePage(path, returned || t.nullLiteral(), componentType, refId, wxses, isApp)
        ast.program.body.push(
          classDecl,
          t.exportDefaultDeclaration(t.identifier(componentType !== 'App' ? defaultClassName : 'App'))
        )
        // path.insertAfter(t.exportDefaultDeclaration(t.identifier(defaultClassName)))
        path.remove()
      }
    },
    VariableDeclaration (path) {
      const variableDeclarator = path.node.declarations[0]
      variableDeclarator.init
      if (t.isVariableDeclarator(variableDeclarator)) {
        // 去除commonjs转es6生成的var module = {exports: {},}
        if (
          t.isIdentifier(variableDeclarator.id) &&
          variableDeclarator.id.name === 'module' &&
          t.isObjectExpression(variableDeclarator.init)
        ) {
          const init: t.ObjectExpression = variableDeclarator.init
          if (
            init.properties &&
            init.properties.length > 0 &&
            t.isObjectProperty(init.properties[0]) &&
            t.isIdentifier(init.properties[0].key) &&
            init.properties[0].key.name === 'exports'
          ) {
            path.remove()
          }
        }
        // 去除commonjs转es6生成的var exports = module.exports
        if (
          t.isIdentifier(variableDeclarator.id) &&
          variableDeclarator.id.name === 'exports' &&
          t.isMemberExpression(variableDeclarator.init)
        ) {
          const init: t.MemberExpression = variableDeclarator.init
          if (
            t.isIdentifier(init.object) &&
            init.object.name === 'module' &&
            t.isIdentifier(init.property) &&
            init.property.name === 'exports'
          ) {
            path.remove()
          }
        }
      }
    },
    ExportDefaultDeclaration (path) {
      const declaration = path.node.declaration
      // 去除commonjs转es6生成的export default module.exports
      if (t.isMemberExpression(declaration)) {
        if (
          t.isIdentifier(declaration.object) &&
          declaration.object.name === 'module' &&
          t.isIdentifier(declaration.property) &&
          declaration.property.name === 'exports'
        ) {
          path.remove()
        }
      }
    },
  }

  traverse(ast, vistor)

  if (!foundWXInstance) {
    ast = parseCode(script + ';Component({})')
    traverse(ast, vistor)
  }

  const taroComponentsImport = buildImportStatement('@tarojs/components', [...usedComponents])
  const taroImport = buildImportStatement('@tarojs/taro', [], 'Taro')
  const reactImport = buildImportStatement('react', [], 'React')
  const withWeappImport = buildImportStatement('@tarojs/with-weapp', [], 'withWeapp')
  ast.program.body.unshift(
    taroComponentsImport,
    reactImport,
    taroImport,
    withWeappImport,
    ...wxses.filter((wxs) => !wxs.src.startsWith('./wxs__')).map((wxs) => buildImportStatement(wxs.src, [], wxs.module))
  )

  return ast
}

function parsePage (
  pagePath: NodePath<t.CallExpression>,
  returned: t.Expression,
  componentType?: string,
  refId?: Set<string>,
  wxses?: WXS[],
  isApp = false
) {
  const stateKeys: string[] = []
  pagePath.traverse({
    CallExpression (path) {
      const callee = path.get('callee')
      replaceIdentifier(callee as NodePath<t.Node>)
      replaceMemberExpression(callee as NodePath<t.Node>)
    },
  })
  if (refId) {
    refId.forEach((id) => {
      if (!stateKeys.includes(id)) {
        stateKeys.push(id)
      }
    })
  }
  const propsKeys: string[] = []
  const arg = pagePath.get('arguments')[0]

  const classBody: any[] = []
  if (arg.isObjectExpression() || arg.isIdentifier()) {
    //
  } else {
    throw codeFrameError(arg.node, `${componentType || '组件'} 的第一个参数必须是一个对象或变量才能转换。`)
  }

  const wxsNames = new Set(wxses ? wxses.map((w) => w.module) : [])

  const renderFunc = buildRender(
    componentType === 'App'
      ? t.memberExpression(t.memberExpression(t.thisExpression(), t.identifier('props')), t.identifier('children'))
      : returned,
    stateKeys.filter((s) => !wxsNames.has(s)),
    propsKeys
  )

  const classDecl = t.classDeclaration(
    t.identifier(componentType === 'App' ? 'App' : defaultClassName),
    t.memberExpression(t.identifier('React'), t.identifier('Component')),
    t.classBody(classBody.concat(renderFunc)),
    []
  )

  classDecl.decorators = [buildDecorator(arg.node, isApp)]

  return classDecl
}
