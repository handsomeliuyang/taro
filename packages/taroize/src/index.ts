import * as t from '@babel/types'
import { promoteRelativePath } from '@tarojs/helper'
import * as path from 'path'

import { errors, resetGlobals, THIRD_PARTY_COMPONENTS } from './global'
import { parseScript } from './script'
import { setting } from './utils'
import { parseVue } from './vue'
import { parseWXML } from './wxml'

interface Option {
  json?: string
  script?: string
  scriptPath?: string
  wxml?: string
  path: string
  rootPath: string
  framework: 'react' | 'vue'
  isApp?: boolean
}

export function parse (option: Option) {
  resetGlobals()
  setting.rootPath = option.rootPath
  if (option.json) {
    const config = JSON.parse(option.json)
    const usingComponents = config.usingComponents
    if (usingComponents) {
      for (const key in usingComponents) {
        if (usingComponents.hasOwnProperty(key)) {
          THIRD_PARTY_COMPONENTS.add(key)
        }
      }
    }
  }

  if (option.framework === 'vue') {
    const result = parseVue(option.path, option.wxml || '', option.script)
    return {
      ...result,
      errors,
    }
  }

  const { wxml, wxses, imports, refIds } = parseWXML(option.path, option.wxml)
  setting.sourceCode = option.script!
  const ast = parseScript(option.script, option.scriptPath, wxml as t.Expression, wxses, refIds, option.isApp)

  // 添加 cacheOptions 引入语句
  if ( option.scriptPath ) {
    const currentPath = option.scriptPath || ''
    const cacheOptionsPath = path.resolve(option.rootPath, 'utils', '_cacheOptions.js')
    const importCacheOptionsUrl = promoteRelativePath(path.relative(currentPath, cacheOptionsPath))
    const importCacheOptionsAst = t.importDeclaration([
      t.importSpecifier(t.identifier('cacheOptions'), t.identifier('cacheOptions'))
    ], t.stringLiteral(importCacheOptionsUrl))
    ast.program.body.unshift(importCacheOptionsAst)
  }

  return {
    ast,
    imports,
    errors,
  }
}
