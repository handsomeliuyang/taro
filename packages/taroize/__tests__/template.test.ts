import * as fs from 'fs-extra'

import { globals } from '../src/global'
import { parse } from '../src/index'
import { getSrcRelPath } from '../src/template'
import { parseWXML } from '../src/wxml'
import { generateMinimalEscapeCode, removeBackslashesSerializer } from './util'

const path = require('path')

const option: any = {
  framework: 'react',
  json: '{}',
  logFilePath: '',
  path: '',
  rootPath: '',
  script: '',
  scriptPath: '',
  wxml: '',
  isApp: false,
}

expect.addSnapshotSerializer(removeBackslashesSerializer)

jest.mock('fs', () => ({
  ...jest.requireActual('fs'), // 保留原始的其他函数
  appendFile: jest.fn(),
}))

describe('template.ts', () => {
  describe('模板', () => {
    // 还原模拟函数
    afterEach(() => {
      jest.restoreAllMocks()
    })

    test('直接使用页面中的模板，转换后会创建模板文件', () => {
      option.wxml = `
        <template name="msgItem">
          <view>
            <text> {{index}}: {{msg}} </text>
            <text> Time: {{time}} </text>
          </view>
        </template>
        <template is="msgItem" data="{{...item}}"/>
      `
      option.script = `
        Page({
          data:{
            item: {
              index: 0,
              msg: 'this is a template',
              time: '2016-09-15'
            }
          },
        })
      `
      option.path = 'template'
      const { ast, imports }: any = parse(option)
      // 模板的引入需在 taro-cli-convertor 验证
      const code = generateMinimalEscapeCode(ast)
      const importsCode = generateMinimalEscapeCode(imports[0].ast)
      expect(code).toMatchSnapshot()
      expect(importsCode).toMatchSnapshot()
    })

    test('当template同时使用外部工具类和data传递数据', () => {
      const wxml = `
      <wxs src="../utils/myFunc.wxs" module="myFunc"/>
      <wxs src="../utils/timFunc.wxs" module="Tim"/>
      <wxs src="../utils/timFunc.wxs" module="utils"/>
      <template name="huangye">
        <text>{{ myFunc.getMsg }}</text>
        <view>{{ Tim.getMsg }}</view>
        <view>{{ '姓名：' + info.name + '年龄：' + info.age }}</view>
      </template>
      <template is="huangye" data="{{ info }}"/>
      `
      // 确定解析wxml文件的绝对路径
      // const dirPath = '\\wechatTest\\template_test\\components\\LunaComponent\\ListHuangye'
      const dirPath = path.join('wechatTest', 'template_test', 'components', 'LunaComponent', 'ListHuangye')
      const rootPathMock = path.join('wechatTest','template_test')
      
      /**
       *  模拟全局对象下的文件路径
       */
      // 保存原始的属性描述符
      const originalDescriptor = Object.getOwnPropertyDescriptor(globals, 'rootPath')
      Object.defineProperty(globals, 'rootPath', {
        get: jest.fn().mockReturnValue(rootPathMock),
      })

      const { imports } = parseWXML(dirPath, wxml)
      const importsCode = generateMinimalEscapeCode(imports[0].ast)
      expect(importsCode).toMatchSnapshot()

      // 恢复原始的属性描述符
      if (originalDescriptor) {
        Object.defineProperty(globals, 'rootPath', originalDescriptor)
      }
    })
  })

  describe('引入', () => {
    describe('import 正常情况', () => {
      // 还原模拟函数
      afterEach(() => {
        jest.restoreAllMocks()
      })

      test('import 引入template', () => {
        const wxmlStr = `
          <import src="../template/template"/>
          <view>
            <template is="template_demo"></template>
          </view>
        `
        const template = `
          <template name="template_demo">
            <view>模版DEMO</view>
          </template>
        `
        // 将 template.ts 中的 fs.readFileSync 返回值模拟为常量 template
        jest.spyOn(fs, 'readFileSync').mockReturnValue(template)
        const dirPath = 'import_normal'
        const { wxml, imports }: any = parseWXML(dirPath, wxmlStr)
        const wxmlCode = generateMinimalEscapeCode(wxml)
        const importsCode = generateMinimalEscapeCode(imports[0].ast)
        expect(wxmlCode).toBe(`<View><TemplateDemoTmpl></TemplateDemoTmpl></View>`)
        expect(importsCode).toMatchSnapshot()
      })

      test('import src 为绝对路径且导入文件不存在', () => {
        const wxmlStr = `
          <import src="/pages/template/template"/>
          <template is="template_demo"></template>
        `
        const template = `
          <template name="template_demo">
            <view>模版DEMO</view>
          </template>
        `
        jest.spyOn(path, 'join').mockReturnValue('\\code\\taro_demo\\pages\\template\\template')
        jest.spyOn(path, 'relative').mockReturnValue('../template/template')
        jest.spyOn(fs, 'readFileSync').mockReturnValue(template)
        jest.spyOn(fs, 'existsSync').mockReturnValue(true)
        const dirPath = 'import_absoulte_path'
        const { wxml, imports }: any = parseWXML(dirPath, wxmlStr)
        const wxmlCode = generateMinimalEscapeCode(wxml)
        const importsCode = generateMinimalEscapeCode(imports[0].ast)
        expect(wxmlCode).toBe('<TemplateDemoTmpl></TemplateDemoTmpl>')
        expect(importsCode).toMatchSnapshot()
      })

      test('import src 为绝对路径但导入文件不存在', () => {
    
        jest.spyOn(path, 'join').mockReturnValue('\\code\\taro_demo\\pages\\template\\myTmpl')
    
        const dirPath = 'import_absoulte_path'
        const srcValue = '/pages/template/myTmpl'
        expect(() => getSrcRelPath(dirPath, srcValue)).toThrowError(`import/include 的 src 请填入正确路径再进行转换：src="${srcValue}"`)
      })
    })

    describe('import 异常情况', () => {
      // 还原模拟函数
      afterEach(() => {
        jest.restoreAllMocks()
      })

      test('当 import 没有 src', () => {
        const wxml = `<import/>`
        const dirPath = 'import_no_src'
        expect(() => parseWXML(dirPath, wxml)).toThrowError('import 标签必须包含 `src` 属性')
      })
    })

    describe('include 正常情况', () => {
      // 还原模拟函数
      afterEach(() => {
        jest.restoreAllMocks()
      })

      test('include 引入 header文件', () => {
        const header = '<text>header.wxml</text>'
        option.wxml = `<include src="../header/header"></include>`
        option.script = 'Page({})'
        option.path = 'include'
        jest.spyOn(fs, 'readFileSync').mockReturnValue(header)
        const { ast }: any = parse(option)
        const wxmlCode = generateMinimalEscapeCode(ast)
        expect(wxmlCode).toMatchSnapshot()
      })
    })

    describe('include 异常情况', () => {
      // 还原模拟函数
      afterEach(() => {
        jest.restoreAllMocks()
      })

      test('include 没有src', () => {
        const wxml = `<include/>`
        const dirPath = 'include_no_src'
        expect(() => parseWXML(dirPath, wxml)).toThrowError('include 标签必须包含 `src` 属性')
      })
    })
  })

  // 主要测试template相关信息（所需变量、要传递的方法）的抽取
  describe('template套用', () => {
    test('同界面template套用', () => {
      const wxmlStr = `
        <template is="firstTag" data='{{ tagListFirst, tagListSecond}}'>顶层</template>

        <template name="firstTag">                   
          <block wx:for="{{tagListFirst}}" wx:key="index">
            <view catchtap="onClickFirstTag">{{item}}</view>
          </block>
          <template is="secondTag" data="{{tagListSecond}}"></template>
        </template>

        <template name="secondTag">
          <block wx:for="{{tagListSecond}}" wx:key="index">
            <view catchtap="onClickSecndTag">{{item}}</view>
          </block>
        </template>
      `
      const dirPath = 'template_apply'
      const { wxml, imports }: any = parseWXML(dirPath, wxmlStr)
      const wxmlCode = generateMinimalEscapeCode(wxml)
      expect(wxmlCode).toMatchSnapshot()
      imports.forEach((importWxml) => {
        const importCode = generateMinimalEscapeCode(importWxml.ast)
        expect(importCode).toMatchSnapshot()
      })
    })

    test('通过import的套用', () => {
      const wxmlStr = `
        <import src="/template/tmplA/index"></import>
        <template is="tmplA" data="{{motto, mott}}"></template>
      `
      const tmplA = `
        <import src="/template/tmplB/index"></import>
        <template name="tmplA">
          <view>this is template A</view>
          <view wx:if="{{motto}}">{{mott}}</view>
          <template is="tmplB"></template>
        </template>
      `
      const tmplB = `
        <template name="tmplB">
          <view>this is template B</view>
          <button bind:tap="onClickC">模板C的按钮</button>
        </template>
      `
      
      jest.spyOn(fs, 'readFileSync').mockReturnValueOnce(tmplA).mockReturnValueOnce(tmplB)
      jest.spyOn(fs, 'existsSync').mockReturnValue(true)
      const dirPath = 'template_import_apply'
      const { wxml, imports }: any = parseWXML(dirPath, wxmlStr)
      const wxmlCode = generateMinimalEscapeCode(wxml)
      expect(wxmlCode).toMatchSnapshot()
      imports.forEach((importWxml) => {
        const importCode = generateMinimalEscapeCode(importWxml.ast)
        expect(importCode).toMatchSnapshot()
      })
    })
  })
})
