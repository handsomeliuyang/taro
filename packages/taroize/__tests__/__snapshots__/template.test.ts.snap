// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template.ts 引入 import 正常情况 import src 为绝对路径 1`] = `
"import { Block, View } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
import withWeapp from \\"@tarojs/with-weapp\\";
@withWeapp({})
class TemplateDemoTmpl extends React.Component {
  render() {
    return <Block><View>模版DEMO</View></Block>;
  }
}
export default TemplateDemoTmpl;"
`;

exports[`template.ts 引入 import 正常情况 import 引入template 1`] = `
"import { Block, View } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
import withWeapp from \\"@tarojs/with-weapp\\";
@withWeapp({})
class TemplateDemoTmpl extends React.Component {
  render() {
    return <Block><View>模版DEMO</View></Block>;
  }
}
export default TemplateDemoTmpl;"
`;

exports[`template.ts 引入 include 正常情况 include 引入 header文件 1`] = `
"import { Block, Text } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
Page({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <Text>header.wxml</Text>;
  }
}
export default _C;"
`;

exports[`template.ts 模板 当template同时使用外部工具类和data传递数据 1`] = `
"import { Block, Text, View } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
import withWeapp from \\"@tarojs/with-weapp\\";
import myFunc from \\"../components/LunaComponent/utils/myFunc.wxs.js\\";
import Tim from \\"../components/LunaComponent/utils/timFunc.wxs.js\\";
@withWeapp({})
class HuangyeTmpl extends React.Component {
  render() {
    const {
      data: {
        info
      }
    } = this.props;
    return <Block><Text>{myFunc.getMsg}</Text><View>{Tim.getMsg}</View><View>{'姓名：' + info.name + '年龄：' + info.age}</View></Block>;
  }
}
export default HuangyeTmpl;"
`;

exports[`template.ts 模板 直接使用页面中的模板，转换后会创建模板文件 1`] = `
"import { Block, View, Text } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
const {
  default: withWeapp
} = require(\\"@tarojs/with-weapp\\");
Page({
  data: {
    item: {
      index: 0,
      msg: 'this is a template',
      time: '2016-09-15'
    }
  }
});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      item
    } = this.data;
    return <MsgItemTmpl data={item}></MsgItemTmpl>;
  }
}
export default _C;"
`;

exports[`template.ts 模板 直接使用页面中的模板，转换后会创建模板文件 2`] = `
"import { Block, View, Text } from \\"@tarojs/components\\";
import React from \\"react\\";
import Taro from \\"@tarojs/taro\\";
import withWeapp from \\"@tarojs/with-weapp\\";
@withWeapp({})
class MsgItemTmpl extends React.Component {
  render() {
    const {
      data: {
        index,
        msg,
        time
      }
    } = this.props;
    return <Block><View><Text>{index + \\": \\" + msg}</Text><Text>{\\"Time: \\" + time}</Text></View></Block>;
  }
}
export default MsgItemTmpl;"
`;
